pyteal/ast/frame.py

```diff
@@ -0,0 +1,192 @@
+from typing import TYPE_CHECKING
+
+from pyteal.ast.expr import Expr
+from pyteal.types import TealType, require_type
+from pyteal.errors import TealInputError, verifyProgramVersion
+from pyteal.ir import TealBlock, TealSimpleBlock, TealOp, Op
+
+if TYPE_CHECKING:
+    from pyteal.compiler import CompileOptions
+
+
+class Proto(Expr):
+    def __init__(self, num_args: int, num_returns: int):
+        super().__init__()
+        if num_args < 0:
+            raise TealInputError(
+                f"the number of arguments provided to Proto must be >= 0 but {num_args=}"
+            )
+        if num_returns < 0:
+            raise TealInputError(
+                f"the number of return values provided to Proto must be >= 0 but {num_returns=}"
+            )
+        self.num_args = num_args
+        self.num_returns = num_returns
+
+    def __teal__(self, options: "CompileOptions") -> tuple[TealBlock, TealSimpleBlock]:
+        verifyProgramVersion(
+            Op.proto.min_version,
+            options.version,
+            "Program version too low to use op proto",
+        )
+        op = TealOp(self, Op.proto, self.num_args, self.num_returns)
+        return TealBlock.FromOp(options, op)
+
+    def __str__(self) -> str:
+        return f"(proto: num_args = {self.num_args}, num_returns = {self.num_returns})"
+
+    def type_of(self) -> TealType:
+        return TealType.none
+
+    def has_return(self) -> bool:
+        return False
+
+
+Proto.__module__ = "pyteal"
+
+
+class FrameDig(Expr):
+    def __init__(self, depth: int):
```

Comment:
According to [Martin Fowler and Phil Karlton](https://martinfowler.com/bliki/TwoHardThings.html):

> There are only two hard things in Computer Science: cache invalidation and naming things.

We've seen how hard the first problem is in our recent boxes release. The second problem occurs more often.

So what should we name `depth`? There's a lot to argue for keeping this name as is. I'll leave the positives for the counter-arguments. But here are what to my ear strikes me as non-ideal about "depth".
* Usually, depth is an unsigned value (though here's a [counter-example](https://www.usgs.gov/observatories/hvo/news/volcano-watch-why-do-some-earthquakes-have-negative-depths) )
* Using at in a natural way, conforms with the volcano link above. So that positive depth -to me- would mean going _below_ the frame pointer into the args, and negative would mean going above, to the return values. This is the opposite of how we would be using it
* There is already an upcoming example where we provide a different name for the argument: [frame_index](https://github.com/algorand/pyteal/blob/6ba772a69721fe43409a7fa03f8235741dd38d8b/pyteal/ast/subroutine.py#L975)

So I propose we rename the parameter to one of:
* `frame_index`
* `displacement`
* `index`
* `delta`
* something else?

Reply:
BTW: `height` is off the table, as it already means the height of the frame pointer.

Reply:
`depth` still seems like a good name for the argument of the non frame `Bury` and `Dig`

Reply:
yea I agree, any of proposed naming should be more pertinent than `depth` for it suggests a number >= 0.

I would rename it to `frame_index`, feels most making sense to me.